### PR TITLE
Display images on tensorboard and add parameters

### PR DESCRIPTION
**Pull request by Shiro-LK**
I add the possibility to display images on tensorboard, and added some parameters : choice of optimizers and learning rate, use of dataset dir for CSV format, compute the validation loss possible during training.

### CODE REVIEW

File: keras_retinanet/bin/train.py [hunk 1]
```diff
@@ -352,6 +354,25 @@ def check_args(parsed_args):
 
     return parsed_args
 
+def optimizers(args):
```
Comment (hgaiser) [inline keras_retinanet/bin/train.py]: Prefer to name it `create_optimizer`, so that it is in line with the other names.

File: keras_retinanet/bin/train.py [hunk 2]
```diff
@@ -352,6 +354,25 @@ def check_args(parsed_args):
...
+        opt = Adagrad(lr=args.lr, clipnorm=args.clip_norm)
+    elif args.optimizer == 'nadam':
+        opt = Nadam(lr=args.lr, clipnorm=args.clip_norm)
+    elif args.optimizer == 'rmsprop':
+        opt = RMSprop(lr=args.lr, clipnorm=args.clip_norm)
+        
+    else:
+        opt = Adam(lr=args.lr, clipnorm=args.clip_norm)
+        
+    print(args.optimizer + ':' + str(args.lr) + ' ' + str(args.clip_norm))
```
Comment (hgaiser) [inline keras_retinanet/bin/train.py]: `print('{}: {} {}'.format(args.optimizer, args.lr, args.clip_norm)`

File: keras_retinanet/bin/train.py [hunk 3]
```diff
@@ -406,6 +428,12 @@ def csv_list(string):
     parser.add_argument('--image-min-side', help='Rescale the image so the smallest side is min_side.', type=int, default=800)
     parser.add_argument('--image-max-side', help='Rescale the image if the largest side is larger than max_side.', type=int, default=1333)
 
+    parser.add_argument('--tensorboxes', dest='tensorboxes',  help='Number of images to store on tensorboard', type=int, default=0)
```
Comment (hgaiser) [inline keras_retinanet/bin/train.py]: The `dest` is already `tensorboxes` by default, why assign it? (same for `optimizer`, `lr`, `clip_norm`).

File: keras_retinanet/bin/train.py [hunk 4]
```diff
@@ -406,6 +428,12 @@ def csv_list(string):
...
+    parser.add_argument('--use-val-loss', dest='val_loss', help='Compute validation loss', type=bool, default=False)
```
Comment (hgaiser) [inline keras_retinanet/bin/train.py]: If you use `action='store_true'`, then `type` and `bool` aren't necessary.

File: keras_retinanet/bin/train.py [hunk 5]
```diff
@@ -406,6 +428,12 @@ def csv_list(string):
...
+    parser.add_argument('--optimizer', dest="optimizer", help='choose optimizer : sgd, adam, rmsprop, adagrad, adadelta, adamax, nadam. default=adam', default='adam')
```
Comment (hgaiser) [inline keras_retinanet/bin/train.py]: `help='Optimizer to use (one of sgd, adam (default), rmsprop, adagrad, adadelta, adamax, nadam).'`

also, maybe use the `choices` parameter for `add_argument`.

File: keras_retinanet/bin/train.py [hunk 6]
```diff
@@ -415,6 +443,8 @@ def main(args=None):
         args = sys.argv[1:]
     args = parse_args(args)
 
+    writer = tf.summary.FileWriter(args.tensorboard_dir+'/image')
```
Comment (hgaiser) [inline keras_retinanet/bin/train.py]: You can't use `tf` here, we only import it when we choose to use tensorboard.

Also, it's missing spaces around the `+` operator.

File: keras_retinanet/bin/train.py [hunk 7]
```diff
@@ -467,16 +498,34 @@ def main(args=None):
         prediction_model,
         validation_generator,
         args,
+        tensorboard_image=writer,
     )
 
     # start training
-    training_model.fit_generator(
+    # start training
```
Comment (hgaiser) [inline keras_retinanet/bin/train.py]: double comment

File: keras_retinanet/bin/train.py [hunk 8]
```diff
@@ -467,16 +498,34 @@ def main(args=None):
...
+    if args.val_loss:
+        
+        training_model.fit_generator(
+            generator=train_generator,
+            steps_per_epoch=args.steps,
+            validation_data=validation_generator,
```
Comment (hgaiser) [inline keras_retinanet/bin/train.py]: I think this can be shortened by doing:

`validation_steps=validation_generator if args.val_loss else None`

Then the `if` is not required.

File: keras_retinanet/callbacks/eval.py [hunk 9]
```diff
@@ -54,20 +55,13 @@ def on_epoch_end(self, epoch, logs=None):
...
             score_threshold=self.score_threshold,
             max_detections=self.max_detections,
-            save_path=self.save_path
+            save_path=self.save_path,
+            writer=self.tensorboard_image,
+            steps=epoch,
+            number= self.number_images,
         )
 
-        # compute per class average precision
```
Comment (hgaiser) [inline keras_retinanet/callbacks/eval.py]: You undid this commit.

File: keras_retinanet/utils/eval.py [hunk 10]
```diff
@@ -21,9 +21,45 @@
 
 import numpy as np
 import os
-
+import io
+import tensorflow as tf
```
Comment (hgaiser) [inline keras_retinanet/utils/eval.py]: `eval.py` shouldn't depend on tf, only in the paths where tf is actually required

File: keras_retinanet/utils/eval.py [hunk 11]
```diff
@@ -21,9 +21,45 @@
...
+from PIL import Image
 import cv2
 
+def TensorboardImage(writer, image, number=0, step=0):
```
Comment (hgaiser) [inline keras_retinanet/utils/eval.py]: Why is this in caps?

File: keras_retinanet/utils/eval.py [hunk 12]
```diff
@@ -21,9 +21,45 @@
...
+    '''
+        writer : buffer to write the image
```
Comment (hgaiser) [inline keras_retinanet/utils/eval.py]: These comments are not in line with other comments.

File: keras_retinanet/utils/eval.py [hunk 13]
```diff
@@ -21,9 +21,45 @@
...
+        
+    # temp = images[i,:,:,:]*255
+    temp = np.copy(image)
+    #if channel == 1:
+        #temp = np.expand_dims(temp, axis=-1)
+        
+    temp = temp.astype('uint8')
+    
+        
+    if c == 3:
```
Comment (hgaiser) [inline keras_retinanet/utils/eval.py]: else?

File: keras_retinanet/utils/eval.py [hunk 14]
```diff
@@ -54,7 +90,7 @@ def _compute_ap(recall, precision):
     return ap
 
 
-def _get_detections(generator, model, score_threshold=0.05, max_detections=100, save_path=None):
+def _get_detections(generator, model, score_threshold=0.05, max_detections=100, save_path=None, writer=None, steps=0, number=0):
```
Comment (hgaiser) [inline keras_retinanet/utils/eval.py]: Perhaps it's better to add an argument functor, so that you can execute any function on a predicted image. That way we are hiding the `tf` specific code from this generic function.

File: keras_retinanet/utils/eval.py (hunk 11 above)
Comment (Shiro-LK) [reply keras_retinanet/utils/eval.py]: there is no reason. Do you prefer display_tensorboard_image ?

File: keras_retinanet/utils/eval.py (hunk 10 above)
Comment (Shiro-LK) [reply keras_retinanet/utils/eval.py]: As TensorboardImage call tensorflow function can I create a specific file for this function ?

**Comment by lvaleriu** [comment]
@Shiro-LK Nice PR indeed. Any updates about the requested changes?

**Comment by Shiro-LK** [comment]
@lvaleriu  I correct the code but how can I update the pull request ?

**Comment by lvaleriu** [comment]
By doing a new commit.

File: keras_retinanet/utils/eval.py (hunk 13 above)
Comment (Shiro-LK) [reply keras_retinanet/utils/eval.py]: I do not need else, because for now only channels = 3 is possible, but in the future, if multi channels is added, we will need to add an else part.

**Comment by Shiro-LK** [comment]
I corrected the code

**Comment by Shiro-LK** [comment]
It seems there is a conflict what I have to do ?

**Comment by Axxeption** [comment]
This seems a very interesting feature! I will try to get your branch running, @Shiro-LK

File: .gitignore [hunk 15]
```diff
@@ -2,3 +2,7 @@ __pycache__/
 *.pyc
 /.pytest_cache
 /.cache
+*.h5
+logs*
```
Comment (hgaiser) [inline .gitignore]: Could you make this `logs/*`?

File: .gitignore [hunk 16]
```diff
@@ -2,3 +2,7 @@ __pycache__/
...
+events.out*
```
Comment (hgaiser) [inline .gitignore]: This shouldn't be necessary then?

File: .gitignore [hunk 17]
```diff
@@ -2,3 +2,7 @@ __pycache__/
...
+/snapshot*
```
Comment (hgaiser) [inline .gitignore]: Same as this?

File: keras_retinanet/bin/train.py [hunk 18]
```diff
@@ -352,6 +354,28 @@ def check_args(parsed_args):
...
+        opt = Adagrad(lr=args.lr, clipnorm=args.clip_norm)
+    elif args.optimizer == 'nadam':
+        opt = Nadam(lr=args.lr, clipnorm=args.clip_norm)
+    elif args.optimizer == 'rmsprop':
+        opt = RMSprop(lr=args.lr, clipnorm=args.clip_norm)
+        
+    else:
+        opt = Adam(lr=args.lr, clipnorm=args.clip_norm)
+        
+    print('{}: {} {}'.format(args.optimizer, args.lr, args.clip_norm))
```
Comment (hgaiser) [inline keras_retinanet/bin/train.py]: Remove the print.

File: keras_retinanet/bin/train.py [hunk 19]
```diff
@@ -383,7 +407,8 @@ def csv_list(string):
     csv_parser.add_argument('annotations', help='Path to CSV file containing annotations for training.')
     csv_parser.add_argument('classes', help='Path to a CSV file containing class label mapping.')
     csv_parser.add_argument('--val-annotations', help='Path to CSV file containing annotations for validation (optional).')
-
+    csv_parser.add_argument('--dataset_dir', help='path to the dataset', default=None)
```
Comment (hgaiser) [inline keras_retinanet/bin/train.py]: Please use `--dataset-dir`, it will be converted to an underscore automatically.
Also, please make the help a sentence (start with a capital and end with a period), to keep it in the same style as the rest.

File: keras_retinanet/bin/train.py [hunk 20]
```diff
@@ -406,6 +431,12 @@ def csv_list(string):
     parser.add_argument('--image-min-side', help='Rescale the image so the smallest side is min_side.', type=int, default=800)
     parser.add_argument('--image-max-side', help='Rescale the image if the largest side is larger than max_side.', type=int, default=1333)
 
+    parser.add_argument('--tensorboxes',  help='Number of images to store on tensorboard', type=int, default=0)
+    parser.add_argument('--use-val-loss', dest='val_loss', help='Compute validation loss', action='store_true')
+    parser.add_argument('--optimizer', help='choose optimizer. default=adam', default='adam', choices=['adam', 'sgd', 'adagrad', 'adadelta', 'adamax', 'nadam', 'rmsprop'])
```
Comment (hgaiser) [inline keras_retinanet/bin/train.py]: Same here with the `help` section. Also, I would omit the `choose`, simply "The optimizer to use." will be fine.

File: keras_retinanet/bin/train.py [hunk 21]
```diff
@@ -415,6 +446,8 @@ def main(args=None):
         args = sys.argv[1:]
     args = parse_args(args)
 
+    #writer = tf.summary.FileWriter(args.tensorboard_dir+'/image')
```
Comment (hgaiser) [inline keras_retinanet/bin/train.py]: This can be removed?

File: keras_retinanet/bin/train.py [hunk 22]
```diff
@@ -467,17 +501,24 @@ def main(args=None):
         prediction_model,
         validation_generator,
         args,
+        #tensorboard_image=writer,
```
Comment (hgaiser) [inline keras_retinanet/bin/train.py]: Same with this?

File: keras_retinanet/bin/train.py [hunk 23]
```diff
@@ -467,17 +501,24 @@ def main(args=None):
...
     )
 
     # start training
+
```
Comment (hgaiser) [inline keras_retinanet/bin/train.py]: This can be removed?

File: keras_retinanet/bin/train.py [hunk 24]
```diff
@@ -467,17 +501,24 @@ def main(args=None):
...
+        
     training_model.fit_generator(
         generator=train_generator,
         steps_per_epoch=args.steps,
+        validation_data=validation_generator if args.val_loss else None,
+        validation_steps= int(round(validation_generator.size()/args.batch_size)) if args.val_loss else None,
```
Comment (hgaiser) [inline keras_retinanet/bin/train.py]: This line will probably complain with flake8, because there are no whitespaces around the divide operator.

File: keras_retinanet/bin/train.py [hunk 25]
```diff
@@ -467,17 +501,24 @@ def main(args=None):
...
         epochs=args.epochs,
         verbose=1,
-        callbacks=callbacks,
+        callbacks=callbacks
+
```
Comment (hgaiser) [inline keras_retinanet/bin/train.py]: This can be removed.

File: keras_retinanet/bin/train.py [hunk 26]
```diff
@@ -467,17 +501,24 @@ def main(args=None):
...
+        validation_steps= int(round(validation_generator.size()/args.batch_size)) if args.val_loss else None,
         epochs=args.epochs,
         verbose=1,
-        callbacks=callbacks,
+        callbacks=callbacks
+        
     )
 
 
+
```
Comment (hgaiser) [inline keras_retinanet/bin/train.py]: And this.

File: keras_retinanet/callbacks/eval.py [hunk 27]
```diff
@@ -22,7 +22,7 @@ class Evaluate(keras.callbacks.Callback):
     """ Evaluation callback for arbitrary datasets.
     """
 
-    def __init__(self, generator, iou_threshold=0.5, score_threshold=0.05, max_detections=100, save_path=None, tensorboard=None, verbose=1):
+    def __init__(self, generator, iou_threshold=0.5, score_threshold=0.05, max_detections=100, save_path=None, tensorboard=None, verbose=1, args=None):
```
Comment (hgaiser) [inline keras_retinanet/callbacks/eval.py]: I would prefer not to pass args here, if it requires `tensorboxes`, pass tensorboxes instead.

File: keras_retinanet/callbacks/eval.py [hunk 28]
```diff
@@ -41,7 +41,14 @@ def __init__(self, generator, iou_threshold=0.5, score_threshold=0.05, max_detec
         self.save_path       = save_path
         self.tensorboard     = tensorboard
         self.verbose         = verbose
-
+        
+        self.number_images = args.tensorboxes
+        if self.number_images > 0:
+            import tensorflow as tf
+            self.tensorboard_image= tf.summary.FileWriter(args.tensorboard_dir+'/image')
```
Comment (hgaiser) [inline keras_retinanet/callbacks/eval.py]: This line will also not work due to whitespaces.

File: keras_retinanet/callbacks/eval.py [hunk 29]
```diff
@@ -41,7 +41,14 @@ def __init__(self, generator, iou_threshold=0.5, score_threshold=0.05, max_detec
...
+
```
Comment (hgaiser) [inline keras_retinanet/callbacks/eval.py]: Can be removed.

File: keras_retinanet/callbacks/eval.py [hunk 30]
```diff
@@ -54,20 +61,13 @@ def on_epoch_end(self, epoch, logs=None):
...
-        precision = 0
-        for label, (average_precision, num_annotations ) in average_precisions.items():
-            if self.verbose == 1:
-                print('{:.0f} instances of class'.format(num_annotations),
-                      self.generator.label_to_name(label), 'with average precision: {:.4f}'.format(average_precision))
-            if num_annotations > 0:
-                present_classes += 1
-                precision       += average_precision
-        self.mean_ap = precision / present_classes
+        self.mean_ap = sum(average_precisions.values()) / len(average_precisions)
```
Comment (hgaiser) [inline keras_retinanet/callbacks/eval.py]: I think this changed at some point in `master`, but your commits seem to undo this?

File: keras_retinanet/utils/eval.py [hunk 31]
```diff
@@ -21,10 +21,10 @@
 
 import numpy as np
 import os
-
 import cv2
 
 
+
```
Comment (hgaiser) [inline keras_retinanet/utils/eval.py]: Can be removed.

File: keras_retinanet/utils/eval.py [hunk 32]
```diff
@@ -77,7 +80,7 @@ def _get_detections(generator, model, score_threshold=0.05, max_detections=100,
         image, scale = generator.resize_image(image)
 
         # run network
-        boxes, scores, labels = model.predict_on_batch(np.expand_dims(image, axis=0))[:3]
+        boxes, scores, labels = model.predict_on_batch(np.expand_dims(image, axis=0))
```
Comment (hgaiser) [inline keras_retinanet/utils/eval.py]: Some change from master was undone here.

File: keras_retinanet/utils/eval.py [hunk 33]
```diff
@@ -102,7 +105,19 @@ def _get_detections(generator, model, score_threshold=0.05, max_detections=100,
             draw_detections(raw_image, image_boxes, image_scores, image_labels, label_to_name=generator.label_to_name)
 
             cv2.imwrite(os.path.join(save_path, '{}.png'.format(i)), raw_image)
-
+        else: # training
```
Comment (hgaiser) [inline keras_retinanet/utils/eval.py]: If save_path is None, it doesn't mean we're training.

**Comment by adityapatadia** [comment]
Any update on this?

**Comment by hgaiser** [comment]
Closing due to inactivity. We might pick this up at a later point. Feel free to re-open.
